### PR TITLE
Perf: Improve selection latency by reducing renders

### PR DESCRIPTION
**Pull request by dreamofabear**
## Context

Just more noodling on element selection performance.

## Summary

According to React profiler, reduces cost of text element selection from ~280 to ~100ms. 

- Memoize renders of display elements
- Memoize renders of sections of UI that don't change on element selection: carousel container, header, page menu
- Skip render of contents in design panels when collapsed

Flamegraphs from the first commit following text element selection click:

### Before
Screen Shot 2021-07-12 at 6 40 05 PM

### After
Screen Shot 2021-07-12 at 6 42 31 PM

## Relevant Technical Choices

N/A

## To-do

- Checklist UI components are still rerendering on selection but should be fixed with #8289

## User-facing changes

- Changing element selection should be more responsive

### QA

This PR can be tested by following these steps:

1. Select a text element in the editor
2. The editor should update slightly faster (display selection box & update text panel)
3. The rest of the editor should work normally

### UAT

- [x] UAT should use the same steps as above.

### Does this PR have a security-related impact?

No.

### Does this PR change what data or activity we track or use?

No.

### Does this PR have a legal-related impact?

No.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Partial for #8259

### CODE REVIEW

**Comment by dreamofabear** [comment]
As an aside, I think we should generally discourage unnecessary object instantiation in loops. We saw one surprising example in #8258 but here's a less severe one:

```js
{page.elements.map(({ id, ...rest }) => (
  <DisplayElement
    key={id}
    previewMode
    element={{ id, ...rest }}
    page={page}
  />
))}
```

The instantiation cost is likely minor for smaller arrays, but the cumulative effect across our codebase is probably non-trivial. Also this particular pattern breaks shallow object comparison for APIs like React.memo. 

/cc @barklund @BrittanyIRL

File: assets/src/edit-story/components/panels/panel/shared/content.js [hunk 1]
```diff
@@ -42,13 +42,9 @@ function Content({ children, ...rest }) {
...
   };
 
-  return (
-    <Container
-      style={formStyle}
-      {...rest}
-      id={panelContentId}
-      hidden={isCollapsed}
-    >
+  // Don't render panel content/children when collapsed.
```
Comment (swissspidy) [inline assets/src/edit-story/components/panels/panel/shared/content.js]: Any a11y implications here?

**Comment by swissspidy** [comment]
> As an aside, I think we should generally discourage unnecessary object instantiation in loops. We saw one surprising example in #8258 but here's a less severe one:
> 
> ```js
> {page.elements.map(({ id, ...rest }) => (
>   <DisplayElement
>     key={id}
>     previewMode
>     element={{ id, ...rest }}
>     page={page}
>   />
> ))}
> ```
> 
> The instantiation cost is likely minor for smaller arrays, but the cumulative effect across our codebase is probably non-trivial. Also this particular pattern breaks shallow object comparison for APIs like React.memo.
> 
> /cc @barklund @BrittanyIRL

Is there perhaps an ESLint rule for this?

File: assets/src/edit-story/components/panels/panel/shared/content.js (hunk 1 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/panels/panel/shared/content.js]: @barklund Thoughts

File: assets/src/edit-story/components/panels/panel/shared/content.js [hunk 2]
```diff
@@ -42,13 +42,9 @@ function Content({ children, ...rest }) {
...
+  return isCollapsed ? null : (
+    <Container style={formStyle} {...rest} id={panelContentId}>
       {children}
     </Container>
```
Comment (swissspidy) [inline assets/src/edit-story/components/panels/panel/shared/content.js]: btw this causes some test failures at the moment

**Comment by dreamofabear** [comment]
> btw this causes some test failures at the moment

The integration test is failing because design panels aren't focused after a quick action, which is happening because the node ref passed into `useFocusHighlight()` is null. I think that's because we might be misusing refs and `useEffect()`. [This article](https://medium.com/@teh_builder/ref-objects-inside-useeffect-hooks-eb7c15198780) explains the issue:

> But the ref isn’t updated till after your component has rendered — meaning, any useEffect that skips rendering, won’t see any changes to the ref before the next render pass.

The recommended solution is to use [callback refs](https://reactjs.org/docs/refs-and-the-dom.html#callback-refs).

**Comment by dreamofabear** [comment]
Well this has been a fun little rabbit hole. 

I attempted to convert `useFocusHighlight` and its callers to use callback refs instead of `useRef`, but this has been non-trivial due to code that assumes use of `useRef` e.g. deeply nested forward refs that inspect `ref.current`, providers that store refs, etc. 

A related problem is that `useFocusHighlight` contains `useEffect`, so refactoring it to use callback refs requires callers to use `useState` to re-render components when the ref is set:

```js
function MyComponent() {
  // Will re-render when node is set, boo performance. :(
  const [node, setInput] = useState(null);
  const highlight = useFocusHighlight(key, node);
  return <input ref={setInput} />;
}
```

I'm somewhat convinced now that `useFocusHighlight` should be decomposed to avoid `useEffect` entirely:

```js
function MyComponent() {
  const highlight = useHighlight(state => state[key]));

  const callbackRef = useCallback(node => {
    if (node) {
      // Calls input.focus(), adds event listeners, etc.
      highlightIfNecessary(highlight, node);
    }
  }, [highlight]);

  return <input ref={callbackRef} />;
}
```

This is similar to the pattern [suggested by React here](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node).

**Comment by dreamofabear** [comment]
Moved `useFocusHighlight` changes to #8522 and #8523. This PR will need to wait for those to go in first.

**Comment by dreamofabear** [comment]
This PR is ready for review, PTAL.

**Review by swissspidy — APPROVED** [review]

**Review by miina — APPROVED** [review]

**Comment by swissspidy** [comment]
@googlebot I consent

**Comment by swissspidy** [comment]
Hmm not sure why some media e2e tests are failing now 🤔 
Some Karma tests too apparently.

**Comment by dreamofabear** [comment]
Ah, the e2e tests are failing because the contents of the Accessibility panel (e.g. the "Preview poster image" button) are no longer rendered when the panel is collapsed. Just need to add some code to open the panel first.

**Pull request merged**